keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -1717,45 +1618,76 @@ type ResourceClaimTemplateSpec struct {
 }
 ```
 
-##### Object references
+##### ResourceQuota
+
+At the moment, this type is only used to control whether admin access is
+allowed in a namespace. Other aspects may get added later.
+
+Admin access can be checked in an admission plugin. Other limitations may have
+to be checked at allocation time, either because they depend on device
+attributes which cannot be assumed to be known at the time of admission of a
+claim or because the quota itself influences device selection. For example, a
+future "give me device A if available, otherwise device B" request might not be
+able to provide A because usage of A over quota while device B is still under
+quota.
 
 ```go
-// ResourceClassParametersReference contains enough information to let you
-// locate the parameters for a ResourceClass.
-type ResourceClassParametersReference struct {
-    // APIGroup is the group for the resource being referenced. It is
-    // empty for the core API. This matches the group in the APIVersion
-    // that is used when creating the resources.
-    // +optional
-    APIGroup string
-    // Kind is the type of resource being referenced. This is the same
-    // value as in the parameter object's metadata.
-    Kind string
-    // Name is the name of resource being referenced.
-    Name string
-    // Namespace that contains the referenced resource. Must be empty
-    // for cluster-scoped resources and non-empty for namespaced
-    // resources.
-    // +optional
-    Namespace string
+// Quota controls whether a ResourceClaim may get allocated.
+// Quota is namespaced and applies to claims within the same namespace.
+type Quota struct {
+	metav1.TypeMeta
+	// Standard object metadata.
+	metav1.ObjectMeta
+
+	Spec QuotaSpec
 }
 
-// ResourceClaimParametersReference contains enough information to let you
-// locate the parameters for a ResourceClaim. The object must be in the same
-// namespace as the ResourceClaim.
-type ResourceClaimParametersReference struct {
-    // APIGroup is the group for the resource being referenced. It is
-    // empty for the core API. This matches the group in the APIVersion
-    // that is used when creating the resources.
-    // +optional
-    APIGroup string
-    // Kind is the type of resource being referenced. This is the same
-    // value as in the parameter object's metadata, for example "ConfigMap".
-    Kind string
-    // Name is the name of resource being referenced.
-    Name string
+type QuotaSpec struct {
+	// AllowManagementAccess controls whether claims with ManagementAccess
+	// may be allocated. If multiple quota objects exist and at least one
+	// has a true value, access will be allowed. The default to deny such access.
+	//
+	// +optional
+	AllowManagementAccess bool `json:"allowManagementAccess,omitempty"`
+
+	// Stretch goals for 1.31:
+	//
+	// - maximum number of devices matching a selector
+	// - maximum sum of a certain quantity attribute
```

Comment:
Suggest we just do class and neither of these. The sum is interesting but fraught with challenges (e.g., requested or actually allocated?). Selector is covered more-or-less by class, which also sidesteps the allocation time quota.

Reply:
How do you see quota-on-classes working with partitionable devices?

Reply:
Hmm. I didn't see anything special, what is the issue you are concerned about? This would simply be "you can use this many of this class". The class may or may not allow MIGs.

We probably also need to be able to set the quota to say "no classes except those with specific quota amounts", since by default typically quota is unlimited. In summary: the admin would add a quota object to say "default per-class quota in this namespace is 0", and then add quota entries for each class they want to allow.

Next question is - so far we have talked about quota being on the *devices*. We could simplify and just do quota on the *requests* rather than the *devices resulting from the request*. That could always be done at admission time, though it leaves gaps which the admin would have to satisfy in other ways (this is probably OK; our general quota system is not perfect, either).

On a class-related but off topic concern, I don't think we ever solved your concern with classes that we cannot define a preference in the expressions. For example, "give me non-MIG if it's available, but I will take a MIG". We could handle that with scoring but we're probably not getting there in 1.31. Clayton has also raised the idea of classes with a ranking to ask for "best available". This is sort of like the preferential requests concept we  have discussed but pushed out of 1.31. Probably this topic should move to an issue...

Reply:
I would expect the allocation of a MIG device to consume only a fraction of the quota that would be consumed by a full GPU. I don't see how we can do this using classes alone.

Reply:
It's up to the admin to give more quota for those classes. This is not great, I agree. Better would be to put quota on the underlying resources (e.g., memory). But it does allow some simplistic controls, and is what we can reasonably do in 1.31 (IMO).

Reply:
So I have to use a different class for MIG devices and full GPUs? That seems unfortunate.

Reply:
If the goal is to have a quote mechanism in 1.31, then let's try to pull in a proper solution. These two checks alluded to here are not hard to implement once we agree on the approach.

I can put in more details.

Reply:
See https://github.com/kubernetes/enhancements/blob/b244ddac2a3ec901be513345206f5105c9de3132/keps/sig-node/4381-dra-structured-parameters/README.md?plain=1#L1623-L1676 under `ResourceQuota` and https://github.com/kubernetes/enhancements/pull/4709#pullrequestreview-2110525258

Reply:
> Probably this topic should move to an issue...

https://github.com/kubernetes-sigs/wg-device-management/issues/24

Reply:
I took out b244ddac2a3ec901be5 again and moved that more detailed proposal to the issue.

I would like to discussing during the implementation phase if we can "future proof" the ResourceQuota type, but there's no need to do it now.